evml/keras/models.py

```diff
@@ -1,49 +1,132 @@
-            if num_hidden_layers == 1:
-                model.add(tf.keras.layers.Dense(hidden_size, activation=activation))
-            else:
-                model.add(tf.keras.layers.Dense(hidden_size, activation=activation))
-                model.add(tf.keras.layers.Dropout(dropout_rate))
-      
-    model.add(tf.keras.layers.Dense(output_size, activation=out_activation))
-    
-    return model
+        if self.activation == 'leaky':
+            self.activation = LeakyReLU()
+        
+        if self.kernel_reg == 'l1':
+            self.kernel_reg = L1(self.l1_weight)
+        elif self.kernel_reg == 'l2':
+            self.kernel_reg = L2(self.l2_weight)
+        elif self.kernel_reg == 'l1_l2':
+            self.kernel_reg = L1L2(self.l1_weight, self.l2_weight)
+        else:
+            self.kernel_reg = None
+        
+        for h in range(self.hidden_layers):
+            nn_model = Dense(self.hidden_neurons, activation=self.activation,
+                             kernel_regularizer=L2(self.l2_weight), name=f"dense_{h:02d}")(nn_model)
+            if self.use_dropout:
+                nn_model = Dropout(self.dropout_alpha, name=f"dropout_h_{h:02d}")(nn_model)
+            if self.use_noise:
+                nn_model = GaussianNoise(self.noise_sd, name=f"ganoise_h_{h:02d}")(nn_model)
+        nn_model = DenseNormalGamma(outputs, name='DenseNormalGamma')(nn_model)
+        self.model = Model(nn_input, nn_model)
+        if self.optimizer == "adam":
+            self.optimizer_obj = Adam(lr=self.lr, beta_1=self.adam_beta_1, beta_2=self.adam_beta_2)
+        elif self.optimizer == "sgd":
+            self.optimizer_obj = SGD(lr=self.lr, momentum=self.sgd_momentum)
+        self.model.compile(optimizer=self.optimizer_obj, loss=self.loss,
+                           loss_weights=self.loss_weights, run_eagerly=False)
 
+    def fit(self, x, y):
+        inputs = x.shape[1]
+        if len(y.shape) == 1:
+            outputs = 1
+        else:
+            outputs = y.shape[1]
+        self.build_neural_network(inputs, outputs)
+        self.model.fit(x, y, batch_size=self.batch_size, epochs=self.epochs, verbose=self.verbose, shuffle=True)
+        self.training_var = np.var(y)
+        tf.keras.models.save_model(self.model, self.save_path+'/test_model/model.h5', save_format='h5')
+        return
 
-def calc_prob_uncertainty(outputs, num_classes = 4):
-    evidence = tf.nn.relu(outputs)
-    alpha = evidence + 1
-    S = tf.keras.backend.sum(alpha, axis=1, keepdims=True)
-    u = num_classes / S
-    prob = alpha / S
-    epistemic = prob * (1 - prob) / (S + 1)
-    aleatoric = prob - prob**2 - epistemic
-    return prob, u, aleatoric, epistemic
\ No newline at end of file
+    def predict(self, x, scaler=None):
+        
+        y_out = self.model.predict(x, batch_size=self.batch_size)
+        if self.uncertainties:
+            return self.calc_uncertainties(y_out, scaler)
+        else:
+            return y_out
```

Comment:
For code reading purposes, it is usually preferable to have one return statement per function or method. In this case, you would assign self.calc_uncertainties to a variable called y_out_final if self.uncertainties is True and otherwise assign y_out to y_out_final. Then you would return y_out_final instead of y_out.

Reply:
edited.